hive/src/main/java/org/apache/iceberg/hive/legacy/HiveExpressions.java

```diff
@@ -0,0 +1,271 @@
+   *            partition columns in the simplified expression,
+   *         FALSE if the simplified expression results in an always false expression,
+   *         otherwise returns the simplified expression
+   */
+  static Expression simplifyPartitionFilter(Expression expr, Set<String> partitionColumnNames) {
+    Expression simplified = ExpressionVisitors.visit(expr, new Simplify(partitionColumnNames));
+    if (simplified != null) {
+      // During simplification of IN, NOTIN, NULL, and NOT NULL expressions we introduce additional NOT, TRUE, and FALSE
+      // expressions; so we call Simplify again to remove them
+      simplified = ExpressionVisitors.visit(simplified, new Simplify(partitionColumnNames));
+    }
+    return (simplified == null) ? Expressions.alwaysTrue() : simplified;
+  }
+
+  /**
+   * Converts an {@link Expression} into a filter string which can be passed to the Hive metastore
+   * @param expr The {@link Expression} to be converted into a filter string. This expression must fit the restrictions
+   *             on Hive metastore partition filters. For more details, see {@link Simplify}
+   * @return a filter string equivalent to the given {@link Expression} which can be passed to the Hive metastore
+   */
+  static String toPartitionFilterString(Expression expr) {
+    return ExpressionVisitors.visit(expr, ExpressionToPartitionFilterString.get());
+  }
+
+  /**
+   * Simplifies the {@link Expression} so that it fits the restrictions of the expression that can be passed
+   * to the Hive metastore. It performs the following changes:
+   * 1. Removes any predicates and non-partition columns
+   * 2. Rewrites NOT operators
+   * 3. Removes IS NULL and IS NOT NULL predicates (Replaced with FALSE and TRUE respectively as partition column values
+   *    are always non null for Hive)
+   * 4. Expands IN and NOT IN operators into ORs of EQUAL operations and ANDs of NOT EQUAL operations respectively
+   * 5. Removes any children TRUE and FALSE expressions (Note that the simplified expression still can be TRUE and FALSE
+   *    at the root)
+   */
+  private static class Simplify extends ExpressionVisitors.ExpressionVisitor<Expression> {
+
+    private final Set<String> partitionColumnNamesLowerCase;
+
+    Simplify(Set<String> partitionColumnNames) {
+      this.partitionColumnNamesLowerCase =
+          partitionColumnNames.stream().map(String::toLowerCase).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Expression alwaysTrue() {
+      return Expressions.alwaysTrue();
+    }
+
+    @Override
+    public Expression alwaysFalse() {
+      return Expressions.alwaysFalse();
+    }
+
+    @Override
+    public Expression not(Expression result) {
+      return (result == null) ? null : result.negate();
+    }
+
+    @Override
+    public Expression and(Expression leftResult, Expression rightResult) {
```

Comment:
Would it help if we have a visitor to drop non partition columns  and then use this simplify for the partition columns? Seems like too much is being done here

Reply:
Done